### ISSUE TITLE
[Brainstorm] DonutV2 Ideas?

### ISSUE THREAD

**Issue by logan-markewich** [clovaai/donut#90]
I really like the idea of Donut, It's very fast, doesn't rely on flakey OCR like tesseract, and is very easy to create training data for! 

I've been using Donut exclusively for DocVQA, so that's where my point of view comes from.

After using it for a while, there are a few shortcomings, compared to competitors like LayoutLMV3:
1. (Major) No (easy) way to filter answers. While there are some hacky ways to get the confidence of the sequence generated by the decoder, it's not really consistent enough (correct and wrong answers can basically have any confidence)
2. (Minor) Bounding boxes for answers are missing. Some people have looked into using attention scores to create heatmaps, but this isn't ideal. But in any case, this isn't a totally important feature, it would just make debugging answers easier
3. Probably more!

So I wanted to start a bit of an open brainstorm on ways to improve Donut. Some ideas I had:
1. Train Donut to output "answer not found" or something similar
2. Additional pre-training tasks, in addition to learning to read (maybe something similar to the pre-training tasks in layoutlmv3?)

If anyone else has thoughts, definitely share them! :)

**Comment by satoru814** [comment]
Hi logan 
I follow your first idea!
In OCR system, User want to know the information in documents not outside.
In this points of view, telling "answer not found" is important for user experience.

**Comment by WaterKnight1998** [comment]
> (Major) No (easy) way to filter answers. While there are some hacky ways to get the confidence of the sequence generated by the decoder, it's not really consistent enough (correct and wrong answers can basically have any confidence)

I think this is the major drawback. Could you share those approaches?
Does LayoutLMv3 support Nested fields like ticket rows?

**Comment by logan-markewich** [comment]
@WaterKnight1998 see this thread: https://github.com/clovaai/donut/issues/37

Slightly related, there are some other hacky (although less reliable) ways to get bounding boxes here: https://github.com/clovaai/donut/issues/16

I saw your other post about using donut in production. In my opinion, it's not ready. Even if you train donut to give "answer not found" types of answers, it's not good enough in my experience.

Donut is a cool proof of concept, but for wider adoption, I think it needs some serious changes.

Regarding LayoutLmV3/2, nested fields shoooould work, provided you fine-tune the model on your data appropriately.

**Comment by WaterKnight1998** [comment]
@logan-markewich thank you for sharing the first two resources.

the problem with LayoutLMv3 and LayoutLMv2 is that their license doesn’t allow commercial usage. I tried LayoutLM but I got VRAM memory issues that I didn’t get with LayoutLMv3 or Donut

**Comment by logan-markewich** [comment]
@WaterKnight1998 yup, that is very true. I'm surprised you got VRAM issues though, I think V1 is smaller that V2/V3. Did you use huggingface to test it?

Another one worth checking out is LiLT, which should be better than LayoutLMV1: https://huggingface.co/docs/transformers/model_doc/lilt

**Comment by WaterKnight1998** [comment]
@logan-markewich yes, I used HuggingFace, I thought the same. I will try again on monday.

Yeah, I know LiLT but just 10% better accuracy than V1.

The problem that I found with LayoutModels is that for inference huggingface processor returns more boxes than the ones you input. So it is hard to convert this into a JSON later on. I think this happens if your boxes have several words on it. Have you find this issue before?

**Comment by logan-markewich** [comment]
@WaterKnight1998 10% is still free performance 😉 

I haven't had that issue, but I don't use the processor and do things a little more manually. It returns more boxes probably because tokens are split into sub-tokens. When you get the predicted start/end positions, you have to basically un-tokenize it (using the '##' characters makes this pretty simple in V1/V2) and map things accordingly. 

![image](https://user-images.githubusercontent.com/22285038/205400993-96b4a7fa-7c12-49de-8ba0-5af614bcbe63.png)

**Comment by WaterKnight1998** [comment]
> @WaterKnight1998 10% is still free performance 😉
> 
> I haven't had that issue, but I don't use the processor and do things a little more manually. It returns more boxes probably because tokens are split into sub-tokens. When you get the predicted start/end positions, you have to basically un-tokenize it (using the '##' characters makes this pretty simple in V1/V2) and map things accordingly.
> 
> ![image](https://user-images.githubusercontent.com/22285038/205400993-96b4a7fa-7c12-49de-8ba0-5af614bcbe63.png)

Do you have a notebook showing how to train or predict? All notebook that I see just show boxes in the image. They doesn't convert it into a JSON. 

I would really appreciate if you can share with me one example.

Thanks in advance!

**Comment by logan-markewich** [comment]
@WaterKnight1998 huggingface has a decent notebook for each layoutlm model. In particular, here is an example of training/inference using layoutlmv2 on DocVQA. V1/V3 will be very similar, as well as any QA model that uses boxes 

https://github.com/NielsRogge/Transformers-Tutorials/blob/master/LayoutLMv2/DocVQA/Fine_tuning_LayoutLMv2ForQuestionAnswering_on_DocVQA.ipynb

**Comment by emigomez** [comment]
Your proposals seem very interesting to me! I was trying to obtain the score of each question to know when they are correct or when didn't, but from what I see this value is not very reliable...

Has progress been made in DONUT docvqa to be able to determine if the answer was not found? or is there a way to get a reliable score?

**Comment by Wyzix33** [comment]
How about generating training documents based on a template, i've been using [DocSim](https://github.com/AI4Bharat/DocSim) to generate invoices based on a template with random text and trained donut on the augmented version of them, but this requires some time to prepare the template, maybe we can have a donut version that can recognize the template and generate new images for fine tuning automatically
